### DISCUSSION TITLE
Purge does not remove styles from plugin components

### DISCUSSION THREAD

**Discussion by pfcodes** [tailwindlabs/tailwindcss#6149] [Help]
**What version of Tailwind CSS are you using?**

v2.1.2

**What build tool (or framework if it abstracts the build tool) are you using?**

tailwind-cli (through npx)

**What version of Node.js are you using?**

v17.1.0

**What browser are you using?**

N/A

**What operating system are you using?**

For example: macOS, Windows

**Reproduction URL**

https://play.tailwindcss.com/3NAcPzcdHM?file=config

**Describe your issue**

PurgeCSS does not work on component classes created in the `plugins` part of a base tailwind config used as a preset. This is true even when the requested class to be purged is inside of the PurgeCSS blacklist option.

Command used:
```
npx tailwindcss -c ./tailwind.config.js -i ./input.css -o ./dist/output
```

**Comment by adamwathan** [comment]
I'm guessing this is just because you called the class `button` and the word `button` appears in your HTML (for example even just `<button>` will cause the class to be generated).

Can you create a proper reproduction repository that I can clone and run `npm run build` or whatever to see exactly what you are seeing?

(Going to convert this to a discussion until we've confirmed there's an actual bug, but will convert back if there is)

**Comment by pfcodes** [comment]
@adamwathan 
> I'm guessing this is just because you called the class `button` and the word `button` appears in your HTML (for example even just `<button>` will cause the class to be generated).

That was it. Thank you.

But it'd be nice to still have those elements purged, especially when `mode` is `all` and `preserveHtmlElements` is `false`: 

```
purge: {
  mode: 'all',
  preserveHtmlElements: false,
  content: ['./**/*.{ts,tsx}'],
  options: {
    blacklist: ['button', 'input'],
  },
},
```

Even the [removing all styles](https://tailwindcss.com/docs/optimizing-for-production#removing-all-unused-styles) page emphasizes this configuration _"will really remove **_all_** styles"_.

For now I guess i'll just use different names.

Thanks.

**Reply by adamwathan** [reply]
Agree it would be nice but it's not straightforward since the string 'button' _does_ appear in your files, so from Tailwind's perspective it has to keep any styles named 'button' since there's evidence you might be using them.

This example would generate the `.mt-5` class too for example, since `mt-5` is present in the template:

```html
<div class="text-center">
  Tailwind includes classes like mt-5
</div>
```

Using a very simplistic approach like this is the only way Tailwind can reliably work across all templating languages. And even in the case of the example above, it's possible that you actually are using that instance of `mt-5` as a class:

```html
<div class="text-center">
  Tailwind includes classes like mt-5
</div>

<script>
  let classToAdd = document.querySelector('.text-center).innerText.split(' ')[4].replace('\n', '')
  document.querySelector('.text-center).classList.add(classToAdd)
</script>
```

Silly but demonstrates the point that the best thing for us to do is not be clever about how the file is parsed.

For what it's worth also, all of those extra purge options like `mode`, `preserveHtmlElements`, `blacklist`, etc. do nothing in JIT mode, since we don't actually use PurgeCSS in that mode.

Since `.button` is a custom class, if you don't actually want it in your CSS file why not just delete the class?
